In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import math
import itertools
from matplotlib.ticker import ScalarFormatter
pd.set_option('display.max_columns', None)

In [2]:
df_raw = pd.read_csv("dropout_rate.csv", sep=";", encoding="utf-8-sig")

In [3]:
df_raw

Marital status  Application mode  Application order  Course  \
0                  1                17                  5     171   
1                  1                15                  1    9254   
2                  1                 1                  5    9070   
3                  1                17                  2    9773   
4                  2                39                  1    8014   
...              ...               ...                ...     ...   
4419               1                 1                  6    9773   
4420               1                 1                  2    9773   
4421               1                 1                  1    9500   
4422               1                 1                  1    9147   
4423               1                10                  1    9773   

      Daytime/evening attendance\t  Previous qualification  \
0                                1                       1   
1                                1                       1   
2                                1                       1   
3                                1                       1   
4                                0                       1   
...                            ...                     ...   
4419                             1                       1   
4420                             1                       1   
4421                             1                       1   
4422                             1                       1   
4423                             1                       1   

      Previous qualification (grade)  Nacionality  Mother's qualification  \
0                              122.0            1                      19   
1                              160.0            1                       1   
2                              122.0            1                      37   
3                              122.0            1                      38   
4                              100.0            1                      37   
...                              ...          ...                     ...   
4419                           125.0            1                       1   
4420                           120.0          105                       1   
4421                           154.0            1                      37   
4422                           180.0            1                      37   
4423                           152.0           22                      38   

      Father's qualification  Mother's occupation  Father's occupation  \
0                         12                    5                    9   
1                          3                    3                    3   
2                         37                    9                    9   
3                         37                    5                    3   
4                         38                    9                    9   
...                      ...                  ...                  ...   
4419                       1                    5                    4   
4420                       1                    9                    9   
4421                      37                    9                    9   
4422                      37                    7                    4   
4423                      37                    5                    9   

      Admission grade  Displaced  Educational special needs  Debtor  \
0               127.3          1                          0       0   
1               142.5          1                          0       0   
2               124.8          1                          0       0   
3               119.6          1                          0       0   
4               141.5          0                          0       0   
...               ...        ...                        ...     ...   
4419            122.2          0                          0       0   
4420            119.0          1                  

## Mô tả

Bộ dữ liệu được tạo ra nhằm hỗ trợ giảm tỷ lệ bỏ học và thất bại học tập ở bậc đại học, thông qua việc ứng dụng các kỹ thuật học máy để xác định sinh viên có nguy cơ cao ngay từ giai đoạn đầu của quá trình học. Bộ dữ liệu tạo bởi nhóm Mónica V. Martins, Daniel Tolledo, Jorge Machado, Luís M. T. Baptista, and Valentim Realinho. 2021

In [ ]:
df_raw.shape

In [ ]:
df_raw.isnull().sum()

Số dòng: 4,424

Số cột: 36

Không có giá trị thiếu (ko null).

# Phân loại

## Qualitative
| Cột | Ý nghĩa |
|---|---|
| Marital Status | Tình trạng hôn nhân: 1 – độc thân, 2 – đã kết hôn, 3 – góa, 4 – ly hôn, 5 – ly thân hợp pháp |
| Application mode | Hình thức nộp hồ sơ (ví dụ: kỳ 1, diện đặc biệt, ưu tiên vùng...) |
| Application order | Mức độ ưu tiên của ngành khi đăng ký (0 là ưu tiên cao nhất) |
| Course | Mã ngành học (ví dụ: 33 - Công nghệ nhiên liệu sinh học, 171 - Đồ họa hoạt hình...) |
| Daytime/evening attendance | Thời gian học: 1 – ban ngày, 0 – buổi tối |
| Previous qualification | Trình độ học vấn trước đó |
| Nacionality | Quốc tịch (được mã hóa) |
| Mother's qualification | Trình độ học vấn của mẹ |
| Father's qualification | Trình độ học vấn của cha |
| Mother's occupation | Nghề nghiệp của mẹ |
| Father's occupation | Nghề nghiệp của cha |
| Displaced | Có phải sống xa nhà không: 1 - Có, 0 - Không |
| Educational special needs | Có nhu cầu giáo dục đặc biệt không: 1 - Có, 0 - Không |
| Debtor | Có nợ học phí không: 1 - Có, 0 - Không |
| Tuition fees up to date | Đã đóng đủ học phí chưa: 1 - Rồi, 0 - Chưa |
| Gender | Giới tính: 1 - Nam, 0 - Nữ |
| Scholarship holder | Có học bổng không: 1 - Có, 0 - Không |
| International | Sinh viên quốc tế: 1 - Có, 0 - Không |
| Target | Kết quả học tập: "Graduate" (tốt nghiệp), "Dropout" (bỏ học), ... |

## Quantitative 

### Discrete

| Cột | Ý nghĩa |
|---|---|
| Age at enrollment | Tuổi khi nhập học |
| Curricular units 1st sem (credited) | Số học phần kỳ 1 được miễn/chuyển điểm |
| Curricular units 1st sem (enrolled) | Số học phần kỳ 1 đã đăng ký |
| Curricular units 1st sem (evaluations) | Số học phần kỳ 1 đã tham gia đánh giá |
| Curricular units 1st sem (approved) | Số học phần kỳ 1 đã qua môn |
| Curricular units 1st sem (without evaluations) | Số học phần kỳ 1 không có điểm đánh giá |
| Curricular units 2nd sem (credited) | Số học phần kỳ 2 được miễn/chuyển điểm |
| Curricular units 2nd sem (enrolled) | Số học phần kỳ 2 đã đăng ký |
| Curricular units 2nd sem (evaluations) | Số học phần kỳ 2 đã tham gia đánh giá |
| Curricular units 2nd sem (approved) | Số học phần kỳ 2 đã qua môn |
| Curricular units 2nd sem (without evaluations) | Số học phần kỳ 2 không có điểm đánh giá |

### Continuous

| Cột | Ý nghĩa |
|---|---|
| Previous qualification (grade) | Điểm số bằng cấp trước đó |
| Admission grade | Điểm đầu vào |
| Curricular units 1st sem (grade) | Điểm trung bình kỳ 1 |
| Curricular units 2nd sem (grade) | Điểm trung bình kỳ 2 |
| Unemployment rate | Tỉ lệ thất nghiệp tại thời điểm đó |
| Inflation rate | Tỉ lệ lạm phát tại thời điểm đó |
| GDP | Tổng sản phẩm quốc nội tại thời điểm đó |


In [ ]:
df = df_raw.copy()

summary_stats = {
    "Thông số thống kê": [
        "Mean",
        "Min",
        "25th Percentile",
        "Median",
        "75th Percentile",
        "Max",
        "Mode",
        "Variance",
        "Standard Deviation",
        "Coefficient of Variation"
    ]
}

quantitative_cols = [
    "Age at enrollment",
    "Curricular units 1st sem (credited)",
    "Curricular units 1st sem (enrolled)",
    "Curricular units 1st sem (evaluations)",
    "Curricular units 1st sem (approved)",
    "Curricular units 1st sem (without evaluations)",
    "Curricular units 2nd sem (credited)",
    "Curricular units 2nd sem (enrolled)",
    "Curricular units 2nd sem (evaluations)",
    "Curricular units 2nd sem (approved)",
    "Curricular units 2nd sem (without evaluations)",
    "Previous qualification (grade)",
    "Admission grade",
    "Curricular units 1st sem (grade)",
    "Curricular units 2nd sem (grade)",
    "Unemployment rate",
    "Inflation rate",
    "GDP"
]

def create_summary_df(df, quantitative_cols):
    for col in quantitative_cols:
        data = df[col]
        mean_val = data.mean()
        std_val = data.std()
        mode_val = data.mode().iloc[0] if not data.mode().empty else np.nan
        variance_val = data.var()
        cv_val = std_val / mean_val if mean_val != 0 else np.nan

        summary_stats[col] = [
            mean_val,
            data.min(),
            data.quantile(0.25),
            data.median(),
            data.quantile(0.75),
            data.max(),
            mode_val,
            variance_val,
            std_val,
            cv_val
        ]

    formatted_summary_df = pd.DataFrame(summary_stats).copy()

    for col in formatted_summary_df.columns[1:]:  # Skip "Thông số thống kê"
        formatted_summary_df[col] = formatted_summary_df[col].apply(lambda x: f'{x:,.4f}' if isinstance(x, (int, float)) else x)

    return formatted_summary_df

formatted_summary_df = create_summary_df(df, quantitative_cols)
formatted_summary_df

In [ ]:
def plot_categorical_distribution_grid(df_raw, columns, cols=2):
    rows = (len(columns) + cols - 1) // cols
    fig, axes = plt.subplots(rows, cols, figsize=(6 * cols, 4 * rows))

    # Ensure axes is always a flat array
    if isinstance(axes, plt.Axes):
        axes = np.array([axes])
    else:
        axes = np.array(axes).flatten()

    for i, col in enumerate(columns):
        ax = axes[i]
        counts = df_raw[col].value_counts()
        bars = ax.bar(counts.index, counts.values)

        for bar in bars:
            height = bar.get_height()
            ax.text(bar.get_x() + bar.get_width() / 2, height,
                    f'{int(height)}', ha='center', va='bottom', fontsize=10)

        ax.set_xlabel(col)
        ax.set_ylabel("Count")
        ax.grid(axis='y', linestyle='--', alpha=0.7)
        ax.set_xticks(range(len(counts.index)))
        ax.set_xticklabels(counts.index, rotation=45, ha='right')

    # Remove any unused axes
    for j in range(i + 1, len(axes)):
        fig.delaxes(axes[j])

    plt.tight_layout()
    plt.show()

plot_categorical_distribution_grid(df_raw, ['Target', ], cols=1)

Biểu đồ cho thấy phân bổ dữ liệu không đều giữa ba nhóm Target, với số lượng sinh viên tốt nghiệp (Graduate) chiếm tỷ lệ lớn nhất (~50%), tiếp theo là nhóm bỏ học (Dropout, ~32%), và nhóm đang học (Enrolled) chiếm tỷ lệ thấp nhất (~18%). Áp dụng mô hình phân loại, cần cân nhắc sử dụng các kỹ thuật như resampling (oversampling nhóm thiểu số hoặc undersampling nhóm chiếm đa số), hoặc sử dụng trọng số lớp (class weights) để đảm bảo mô hình không bỏ qua các nhóm quan trọng như Dropout.

In [ ]:
# Số môn đã qua - Năm
df_raw['Curricular units Year'] = df_raw['Curricular units 1st sem (approved)'] + df_raw['Curricular units 2nd sem (approved)']

In [ ]:
sns.histplot(data=df_raw, x='Curricular units Year', hue='Target', kde=True, element="step")

Biểu đồ thể hiện phân phối của tổng số học phần đã hoàn thành trong năm học (Curricular units Year) theo ba nhóm mục tiêu: Dropout, Graduate và Enrolled. Dễ dàng nhận thấy rằng sinh viên thuộc nhóm Dropout chủ yếu tập trung ở mức rất thấp, nhiều người thậm chí không hoàn thành học phần nào hoặc chỉ hoàn thành từ một đến vài học phần, cho thấy họ bỏ học khá sớm. Ngược lại, sinh viên tốt nghiệp (Graduate) có xu hướng hoàn thành đầy đủ số học phần trong năm, với đỉnh phân phối nằm quanh mức 10 đến 12 học phần, phản ánh một quá trình học tập ổn định và đều đặn. Trong khi đó, nhóm Enrolled có phân phối trải rộng hơn, dao động từ thấp đến cao, phù hợp với thực tế rằng nhóm này bao gồm sinh viên đang theo học ở nhiều giai đoạn khác nhau, từ đầu đến cuối chương trình. Tổng thể, biểu đồ cho thấy số lượng học phần hoàn thành trong năm là một đặc trưng có khả năng phân tách rõ giữa các nhóm mục tiêu.

In [ ]:
sns.boxplot(x='Target', y='Curricular units Year', data=df_raw)

Biểu đồ boxplot này mô tả sự phân bố tổng số học phần đã hoàn thành trong năm (Curricular units Year) theo từng nhóm mục tiêu: Dropout, Graduate, và Enrolled. Nhìn chung, sinh viên thuộc nhóm Graduate có giá trị trung vị cao nhất, cho thấy họ thường hoàn thành nhiều học phần hơn so với hai nhóm còn lại. Nhóm Dropout có trung vị thấp nhất và phần lớn dữ liệu nằm ở vùng thấp, thể hiện rằng những sinh viên bỏ học thường hoàn thành rất ít học phần trong năm. Trong khi đó, nhóm Enrolled có trung vị nằm giữa hai nhóm còn lại, phản ánh sự đa dạng về tiến độ học tập của sinh viên đang theo học.

Ngoài ra, cả ba nhóm đều có nhiều điểm ngoại lệ (outliers), đặc biệt là nhóm Graduate với nhiều giá trị vượt xa mức 20 học phần. Điều này cho thấy một số sinh viên có thể đăng ký và hoàn thành số học phần vượt mức bình thường, có thể do học vượt hoặc có chương trình học đặc biệt. Biểu đồ này tiếp tục củng cố nhận định rằng tổng số học phần trong năm là một yếu tố quan trọng giúp phân biệt hành vi học tập giữa các nhóm sinh viên.

In [ ]:
# Điểm tích lũy - Năm
df_raw['Curricular grade Year'] = df_raw['Curricular units 1st sem (grade)'] + df_raw['Curricular units 2nd sem (grade)']

In [ ]:
sns.histplot(data=df_raw, x='Curricular grade Year', hue='Target', kde=True, element="step")

Biểu đồ thể hiện phân phối tổng điểm học phần trong năm học (Curricular grade Year) theo ba nhóm sinh viên: Dropout, Graduate và Enrolled. Nhóm Graduate có phân phối tập trung và đối xứng quanh khoảng 24–28 điểm, cho thấy phần lớn sinh viên tốt nghiệp đạt kết quả học tập tốt, ổn định và nhất quán. Trong khi đó, nhóm Dropout có đặc điểm nổi bật là một lượng lớn sinh viên có tổng điểm bằng 0, thể hiện rõ rệt qua đỉnh nhọn tại giá trị 0 – điều này nhiều khả năng phản ánh việc họ không hoàn thành hoặc không tham gia học tập trong năm học đó.

Ngược lại, nhóm Enrolled có phân phối trải rộng hơn, với phần lớn sinh viên đạt tổng điểm trong khoảng 20–28 điểm, nhưng vẫn có một số phân bố ở mức thấp hơn, phản ánh sự đa dạng trong quá trình học tập khi còn đang theo học. Biểu đồ này cho thấy tổng điểm học phần trong năm là một đặc trưng có giá trị, giúp phân biệt rõ giữa sinh viên bỏ học và sinh viên tốt nghiệp. Nhìn chung, điểm số không chỉ phản ánh năng lực học tập mà còn liên quan chặt chẽ đến khả năng hoàn thành chương trình đào tạo.

In [ ]:
sns.boxplot(x='Target', y='Curricular grade Year', data=df_raw)

Biểu đồ boxplot thể hiện phân phối tổng điểm học phần trong năm học (Curricular grade Year) theo ba nhóm sinh viên: Dropout, Graduate và Enrolled. Nhóm Graduate có giá trị trung vị cao nhất, dao động chủ yếu trong khoảng 25–30 điểm, phản ánh rằng sinh viên tốt nghiệp thường đạt kết quả học tập tốt và ổn định. Nhóm Enrolled có trung vị thấp hơn một chút nhưng vẫn khá gần với nhóm Graduate, cho thấy phần lớn sinh viên đang theo học cũng đạt kết quả học tập khá, mặc dù có nhiều điểm ngoại lệ ở mức điểm thấp – có thể là sinh viên năm đầu hoặc đang gặp khó khăn trong học tập.

Trong khi đó, nhóm Dropout có sự phân tán lớn với trung vị nằm thấp hơn hẳn hai nhóm còn lại, khoảng 12–13 điểm, đồng thời xuất hiện nhiều giá trị rất thấp, thậm chí bằng 0. Điều này cho thấy nhiều sinh viên bỏ học đạt kết quả học tập kém hoặc gần như không tích lũy được điểm số nào trong năm học, là dấu hiệu cảnh báo sớm cho nguy cơ bỏ học. Biểu đồ này một lần nữa khẳng định mối liên hệ chặt chẽ giữa thành tích học tập và khả năng hoàn thành chương trình đào tạo.

In [ ]:
sns.histplot(data=df_raw, x='Age at enrollment', hue='Target', kde=True, element="step")

Biểu đồ thể hiện phân phối độ tuổi nhập học (Age at enrollment) của sinh viên theo ba nhóm mục tiêu: Dropout, Graduate, và Enrolled. Trong đó, nhóm Graduate có độ tuổi nhập học tập trung mạnh ở khoảng 18–20 tuổi, với đỉnh rõ rệt ở tuổi 18. Điều này cho thấy phần lớn sinh viên tốt nghiệp bắt đầu chương trình học đúng độ tuổi truyền thống, ngay sau khi tốt nghiệp phổ thông.

Ngược lại, nhóm Dropout có phân phối trải rộng hơn, với nhiều sinh viên bắt đầu học ở độ tuổi lớn hơn, thậm chí kéo dài đến trên 50 tuổi. Mặc dù nhóm này cũng có đỉnh ở tuổi 18–20, nhưng mật độ thấp hơn so với nhóm Graduate và có nhiều giá trị rải rác ở các độ tuổi lớn hơn. Điều này gợi ý rằng sinh viên bắt đầu học muộn hơn có xu hướng bỏ học cao hơn, có thể do khó khăn trong việc cân bằng học tập với các trách nhiệm khác như gia đình hoặc công việc.

Nhóm Enrolled cũng có xu hướng tập trung ở tuổi 18–21, tương tự như nhóm Graduate, nhưng phân phối vẫn rộng hơn một chút, phản ánh tính đa dạng về độ tuổi của sinh viên đang theo học. Tóm lại, biểu đồ này cho thấy độ tuổi nhập học là một yếu tố có thể liên quan đến khả năng hoàn thành chương trình học, với xu hướng tốt nghiệp cao hơn ở những sinh viên nhập học ở độ tuổi trẻ.

In [ ]:
sns.boxplot(x='Target', y='Age at enrollment', data=df_raw)

Biểu đồ Boxplot thể hiện phân bố độ tuổi khi nhập học theo ba nhóm sinh viên: Dropout (bỏ học), Graduate (tốt nghiệp) và Enrolled (đang học). Từ biểu đồ, có thể nhận thấy sự khác biệt rõ rệt giữa các nhóm, đặc biệt là về độ tuổi trung vị và mức độ phân tán.

Nhóm Dropout có độ tuổi nhập học cao hơn đáng kể so với hai nhóm còn lại. Trung vị của nhóm này rơi vào khoảng 24 tuổi, và dữ liệu có độ phân tán lớn, với nhiều điểm ngoại lệ lên tới 60 hoặc thậm chí 70 tuổi. Điều này cho thấy rằng sinh viên lớn tuổi hơn có xu hướng bỏ học cao hơn, có thể do các yếu tố như áp lực tài chính, gia đình, công việc hoặc khó khăn trong việc quay lại môi trường học tập sau thời gian gián đoạn.

Ngược lại, nhóm Graduate và Enrolled có độ tuổi nhập học thấp và tương đối ổn định, chủ yếu tập trung trong khoảng từ 18 đến 22 tuổi. Trung vị của hai nhóm này gần như trùng nhau, cho thấy phần lớn sinh viên tốt nghiệp hoặc đang tiếp tục học đều là những người nhập học ngay sau khi hoàn thành bậc phổ thông.

In [ ]:
df_raw['Curricular units failed Year'] = df_raw['Curricular units 1st sem (enrolled)'] + df_raw['Curricular units 2nd sem (enrolled)'] - df_raw['Curricular units Year']

In [ ]:
sns.histplot(data=df_raw, x='Curricular units failed Year', hue='Target', kde=True, element="step")


Biểu đồ cho thấy sinh viên tốt nghiệp hầu như không trượt hoặc chỉ trượt rất ít học phần trong năm, với phần lớn tập trung tại giá trị 0–1. Ngược lại, sinh viên bỏ học có số học phần bị trượt trải rộng, nhiều trường hợp trượt trên 5 học phần, cho thấy kết quả học tập yếu kém là yếu tố liên quan trực tiếp đến nguy cơ bỏ học. Nhóm đang theo học nằm giữa hai nhóm còn lại, phản ánh sự đa dạng về năng lực và tiến độ học tập.

In [ ]:
sns.boxplot(x='Target', y='Curricular units failed Year', data=df_raw)

Biểu đồ boxplot cho thấy sinh viên tốt nghiệp thường trượt rất ít học phần trong năm, với trung vị gần bằng 1, trong khi nhóm bỏ học có số học phần trượt cao hơn hẳn, trung vị khoảng 8, và phân bố rộng. Nhóm đang theo học nằm giữa, phản ánh sự phân tán về năng lực học tập. Số học phần bị trượt rõ ràng là một chỉ báo quan trọng liên quan đến nguy cơ bỏ học.

In [ ]:
sns.boxplot(x='Target', y='Admission grade', data=df_raw)

In [ ]:
sns.boxplot(x='Target', y='Unemployment rate', data=df_raw)


In [ ]:
sns.boxplot(x='Target', y='Inflation rate', data=df_raw)

In [ ]:
sns.boxplot(x='Target', y='GDP', data=df_raw)

Không có sự khác biệt đáng kể giữa các nhãn target ở các biểu đồ này

In [ ]:
from scipy import stats

qual_cols = [
    'Marital status', 'Application mode', 'Application order',
    'Daytime/evening attendance\t', 'Previous qualification', 'Nacionality',
    "Mother's qualification", "Father's qualification",
    "Mother's occupation", "Father's occupation", 'Displaced',
    'Educational special needs', 'Debtor', 'Tuition fees up to date',
    'Gender', 'Scholarship holder', 'International'
]

def cramers_v(x, y):
    confusion_matrix = pd.crosstab(x, y)
    chi2 = stats.chi2_contingency(confusion_matrix)[0]
    n = confusion_matrix.sum().sum()
    phi2 = chi2 / n
    r, k = confusion_matrix.shape
    phi2corr = max(0, phi2 - ((k-1)*(r-1))/(n-1))
    rcorr = r - ((r-1)**2)/(n-1)
    kcorr = k - ((k-1)**2)/(n-1)
    return np.sqrt(phi2corr / min((kcorr-1), (rcorr-1)))

cramer_results = {}
for col in qual_cols:
    cramer_results[col] = cramers_v(df_raw[col], df_raw['Target'])

cramer_df = pd.DataFrame.from_dict(cramer_results, orient='index', columns=["Cramér's V"])
cramer_df = cramer_df.sort_values(by="Cramér's V", ascending=False)

plt.figure(figsize=(10, 6))
sns.barplot(x="Cramér's V", y=cramer_df.index, data=cramer_df, palette="viridis")
plt.title("Mức độ tương quan giữa các biến định tính và Target (Cramér's V)")
plt.xlabel("Cramér's V (0 = không liên quan, 1 = tương quan mạnh)")
plt.ylabel("Biến định tính")
plt.tight_layout()
plt.show()

https://en.wikipedia.org/wiki/Cram%C3%A9r%27s_V

Biểu đồ cho thấy một số biến định tính có mối liên hệ đáng kể với biến mục tiêu Target. Đáng chú ý nhất là biến Tuition fees up to date, có hệ số Cramér's V cao nhất (~0.43), cho thấy việc đóng học phí đúng hạn có liên quan mạnh đến khả năng tốt nghiệp hay bỏ học. Tiếp theo là các biến Scholarship holder, Debtor, và Gender, cũng có mức tương quan khá rõ rệt (Cramér's V từ ~0.22–0.28), phản ánh rằng các yếu tố tài chính và giới tính có thể ảnh hưởng đáng kể đến kết quả học tập của sinh viên. Những biến như Nationality, Educational special needs, và International có mức tương quan rất thấp, gần như không có mối liên hệ rõ ràng với kết quả học tập.

In [ ]:

piechart_cols = [
    "Tuition fees up to date",
    "Scholarship holder",
    "Debtor",
    "Gender",
    # "Application mode",
]

for col in piechart_cols:
    counts = df_raw[col].value_counts()
    plt.figure(figsize=(8, 6))
    plt.pie(counts, labels=counts.index, autopct='%1.1f%%', startangle=90)
    plt.title(f'Distribution of {col}')
    plt.axis('equal')  # Equal aspect ratio ensures that pie is drawn as a circle.
    plt.show()

Các biểu đồ tròn cho thấy dữ liệu có sự mất cân đối rõ rệt ở hầu hết các đặc trưng:

Tuition fees up to date và Debtor đều nghiêng mạnh về một phía, với hơn 88% sinh viên đã đóng đủ học phí và không nợ phí, cho thấy phân phối lệch mạnh và mất cân đối giữa hai giá trị 0 và 1.

Scholarship holder cũng bị lệch, với chỉ 24.8% sinh viên nhận học bổng, phản ánh tình trạng phân bố không đều giữa người có và không có học bổng.

Gender tuy có sự chênh lệch (nam chiếm 64.8%), nhưng là đặc trưng có phân phối ít lệch hơn so với các biến còn lại.

Hầu hết các biến dạng nhị phân đều có phân phối mất cân bằng, điều này cần được lưu ý nếu sử dụng các biến này trong mô hình dự đoán để tránh hiện tượng học lệch (bias).

In [ ]:
def plot_heatmap_by_target(df, col1, col2, normalize='index'):
    ct = pd.crosstab(df[col1], df[col2], normalize=normalize)
    plt.figure(figsize=(7, 4))
    sns.heatmap(
        ct,
        annot=True,
        fmt=".2f",
        cmap="YlGnBu",
        cbar_kws={'label': 'Tỷ lệ' if normalize else 'Số lượng'}
    )
    plt.title(f"Heatmap: {col1} vs {col2}")
    plt.xlabel(col2)
    plt.ylabel(col1)
    plt.tight_layout()
    plt.show()

In [ ]:
plot_heatmap_by_target(df_raw, 'Tuition fees up to date', 'Target')

Biểu đồ heatmap cho thấy sinh viên chưa đóng học phí có tỷ lệ bỏ học rất cao (87%), trong khi chỉ 5% trong nhóm này tốt nghiệp. Điều này cho thấy việc đóng học phí đúng hạn có mối liên hệ chặt chẽ với khả năng hoàn thành chương trình học.

In [ ]:
plot_heatmap_by_target(df_raw, 'Scholarship holder', 'Target')

Biểu đồ cho thấy sinh viên có học bổng có tỷ lệ tốt nghiệp rất cao (76%) và tỷ lệ bỏ học thấp (12%). Trong khi đó, sinh viên không có học bổng có tỷ lệ tốt nghiệp và bỏ học gần như ngang nhau (41% và 39%). Điều này gợi ý rằng học bổng có thể là yếu tố hỗ trợ tích cực giúp sinh viên duy trì và hoàn thành chương trình học.


In [ ]:
plot_heatmap_by_target(df_raw, 'Debtor', 'Target')

Biểu đồ heatmap cho thấy sinh viên có nợ (Debtor = 1) có tỷ lệ bỏ học cao (62%) và tỷ lệ tốt nghiệp thấp (20%). Ngược lại, sinh viên không có nợ có tỷ lệ tốt nghiệp cao hơn (54%) và tỷ lệ bỏ học thấp hơn (28%). Điều này cho thấy tình trạng tài chính, thể hiện qua việc nợ học phí, có thể là một yếu tố quan trọng ảnh hưởng đến khả năng hoàn thành chương trình học.


In [ ]:
plot_heatmap_by_target(df_raw, 'Gender', 'Target')

Mức độ chênh lệch không quá lớn và có thể bị ảnh hưởng bởi các yếu tố khác, nên không đủ mạnh để kết luận có tương quan rõ rệt. Đây là một dạng phân bố lệch nhẹ

In [ ]:
plot_heatmap_by_target(df_raw, 'Application mode', 'Target')

Biểu đồ heatmap cho thấy Application mode (phương thức đăng ký tuyển sinh) có ảnh hưởng đáng kể đến kết quả học tập. Một số mã mode như 27, 26, 7, 2 có tỷ lệ bỏ học cực cao (100% hoặc gần 70%), trong khi các mã như 57, 53, 10 lại có tỷ lệ tốt nghiệp rất cao (trên 60%, thậm chí 100%). Điều này cho thấy cách thức nộp đơn đầu vào có thể phản ánh phần nào mức độ sẵn sàng hoặc phù hợp của sinh viên với chương trình học, và nên được cân nhắc kỹ trong quá trình tuyển sinh và hỗ trợ.


## Data transformer

### MinMaxScaler

In [ ]:
# create a copy of df_raw and apply MinMaxScaler to quantitative columns
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

quantitative_cols = quantitative_cols + [
    'Curricular units Year',
    'Curricular grade Year',
    'Curricular units failed Year',
]

df_minmax = df_raw[quantitative_cols + ['Target']].copy()

df_minmax[quantitative_cols] = scaler.fit_transform(df_minmax[quantitative_cols])

df_minmax

In [ ]:
boxplot_cols = [
    'Curricular units Year',
    'Curricular grade Year',
    'Curricular units failed Year',
    'Age at enrollment',
    'Admission grade',
    'Unemployment rate',
    'Inflation rate',
    'GDP',
]

In [ ]:
for col in boxplot_cols:
    plt.figure(figsize=(8, 6))
    sns.boxplot(x='Target', y=col, data=df_minmax)
    plt.title(f'Boxplot of {col} by Target')
    plt.xlabel('Target')
    plt.ylabel(col)
    plt.grid(axis='y', linestyle='--', alpha=0.7)
    plt.tight_layout()
    plt.show()

Biểu đồ dữ liệu một số cột sau khi áp dụng MinMaxScaler

### StandardScaler

In [ ]:
# create a copy of df_raw and apply StandardScaler to quantitative columns

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df_standard = df_raw[quantitative_cols + ['Target']].copy()
df_standard[quantitative_cols] = scaler.fit_transform(df_standard[quantitative_cols])
df_standard

In [ ]:
for col in boxplot_cols:
    plt.figure(figsize=(8, 6))
    sns.boxplot(x='Target', y=col, data=df_standard)
    plt.title(f'Boxplot of {col} by Target')
    plt.xlabel('Target')
    plt.ylabel(col)
    plt.grid(axis='y', linestyle='--', alpha=0.7)
    plt.tight_layout()
    plt.show()

### RobustScaler

In [ ]:
# Copy of df_raw and apply RobustScaler to quantitative columns
from sklearn.preprocessing import RobustScaler
scaler = RobustScaler()
df_robust = df_raw[quantitative_cols + ['Target']].copy()
df_robust[quantitative_cols] = scaler.fit_transform(df_robust[quantitative_cols])
df_robust

In [ ]:
for col in boxplot_cols:
    plt.figure(figsize=(8, 6))
    sns.boxplot(x='Target', y=col, data=df_robust)
    plt.title(f'Boxplot of {col} by Target')
    plt.xlabel('Target')
    plt.ylabel(col)
    plt.grid(axis='y', linestyle='--', alpha=0.7)
    plt.tight_layout()
    plt.show()

### MaxAbsScaler

In [ ]:
# Copy of df_raw and apply MaxAbsScaler to quantitative columns
from sklearn.preprocessing import MaxAbsScaler
scaler = MaxAbsScaler()
df_maxabs = df_raw[quantitative_cols + ['Target']].copy()
df_maxabs[quantitative_cols] = scaler.fit_transform(df_maxabs[quantitative_cols])
df_maxabs


In [ ]:
for col in boxplot_cols:
    plt.figure(figsize=(8, 6))
    sns.boxplot(x='Target', y=col, data=df_maxabs)
    plt.title(f'Boxplot of {col} by Target')
    plt.xlabel('Target')
    plt.ylabel(col)
    plt.grid(axis='y', linestyle='--', alpha=0.7)
    plt.tight_layout()
    plt.show()

### QuantileTransformer

In [ ]:
# Create a copy of df_raw and apply PowerTransformer to quantitative columns
from sklearn.preprocessing import PowerTransformer
scaler = PowerTransformer()
df_power = df_raw[quantitative_cols + ['Target']].copy()
df_power[quantitative_cols] = scaler.fit_transform(df_power[quantitative_cols])
df_power

In [ ]:
for col in boxplot_cols:
    plt.figure(figsize=(8, 6))
    sns.boxplot(x='Target', y=col, data=df_minmax)
    plt.title(f'Boxplot of {col} by Target')
    plt.xlabel('Target')
    plt.ylabel(col)
    plt.grid(axis='y', linestyle='--', alpha=0.7)
    plt.tight_layout()
    plt.show()